index/server/main.go

```diff
@@ -155,22 +159,48 @@ func main() {
 
 		// Serve the index with type
 		if indexType == string(indexSchema.SampleDevfileType) {
-			if iconType == "base64" {
-				bytes, err := encodeIndexIconToBase64(sampleIndexPath)
-				if err != nil {
-					log.Fatal(err.Error())
+			if iconType != "" {
+				if iconType == encodeFormat {
+					if _, err := os.Stat(sampleBase64IndexPath); os.IsNotExist(err) {
+						bytes, err := encodeIndexIconToBase64(sampleIndexPath, sampleBase64IndexPath)
+						if err != nil {
+							c.JSON(http.StatusInternalServerError, gin.H{
+								"status": fmt.Sprintf("failed to encode sample icons to base64 format: %v", err),
+							})
+							return
+						}
+						c.Data(http.StatusOK, http.DetectContentType(bytes), bytes)
+					} else {
+						c.File(sampleBase64IndexPath)
+					}
+				} else {
+					c.JSON(http.StatusBadRequest, gin.H{
+						"status": fmt.Sprintf("the icon type %s is not supported", iconType),
+					})
```

Comment:
imo you can move these logic to a func since its pretty much the same for sample, all and stack. Otherwise lgtm 👍

Reply:
Good point, I refactored it.